### ISSUE TITLE
Places: Cluster View

### ISSUE THREAD

**Issue by red-avtovo** [photoprism/photoprism#1187]
Thank you very much for the project! I really enjoy using it.

I would like to notice, that it would be useful to group photos on a small scale which are very close to each other or let to open cluster view (similar as a moments view) and show the subset of photos from the place, clicked on. It is not very easy to navigate through the subset of the photos or share a group of them from a specific place
![image](https://user-images.githubusercontent.com/10682826/114313988-24ced100-9af9-11eb-984a-9fce0b3271dd.png)

**Comment by shawnbarton** [comment]
I came here to make this exact suggestion. This is an important feature to me so I'm chiming in to show interest. Keep up the great work.

**Comment by lastzero** [comment]
Thank you very much! It's true that we need to optimize this further. As you might have noticed, the maps already cluster automatically - but only up to a certain zoom level as you otherwise might not be able to see individual photos. It's more work to handle situations as this, e.g. by showing a "stack" that then opens in the photo viewer like stacks in regular search results.

**Comment by magnuznilzzon** [comment]
I suspect it is pretty common to have the situation that @shawnbarton has, especially people with families will have a ton of pictures centered on their home. I know I do, that is how I found this thread.

**Comment by el-tiuri** [comment]
Great idea! I like the way it’s implemented in iOS’ Photos app, and this would be pretty similar. The way it works on iOS is that it dynamically creates stacks of photos that are close together relative to the zoom level, and if you tap on a stack you get a grid of all the photos in that stack.

I especially like this because it enables you to easily see all the photos you took in a particular place (regardless of scale - could be a city, could be a continent) regardless of when you took it.

I think limiting the automatic clustering functionality to low zoom levels would be a bit limiting compared to the use it could have, and I don’t really see the benefit of it compared to enabling it on all zoom levels.

Let me know if you want me to share some screenshots of the iOS photos app, in case that’d be helpful.

**Comment by ohthehugemanatee** [comment]
I have the same issue. We in fact already have the dynamic stacks. Just when you click/tap on a circle with a number of photos, that should take you immediately to a pre-filled search results page. That would be plenty.  Users can zoom in closer with the plus/minus keys and get more precise, smaller stacks.

**Comment by fivestones** [comment]
I agree with @ohthehugemanatee. Right now the circles with the number inside just zooms in when clicked. Instead, clicking on these circles should open a search results page with all those photos shown. Zooming in on the map is easily done other ways and doesn't need to be done by clicking on the circles that have numbers inside.

Without the ability to view a search result for a group of photos in a particular location, the maps view is not very useful. For example, it would be really great to be looking at a photo, and easily see all my other photos from the same location or that were taken nearby as a search result. Right now I can go to the location of a photo on the map, but can't just see other photos in the same location without zooming all the way in, and then the other photos are only visible as little circles until I click on each one individually. Another example use would be to just show all the photos taken in, for example, Africa. Or the Grand Canyon. Or Paris. Or my home street. Again, I can find these photos if I look on the map, but only by zooming all the way in to see individual photos and then clicking on each one. If I have a small circle over Africa that says "27", I should be able to just click it and see all those 27 photos taken in Africa in a search result.

**Comment by lastzero** [comment]
To see the photos in regular search results, you can also type the name of the location in the regular search field, for example "Africa". Changing the view type between map and cards / list view causes a lot of load in the front and backend. For this reason, you typically don't want to jump back and forth all the time.

**Comment by fivestones** [comment]
Being able to search for a location and see photos taken there is great. Thanks! I didn't know about that.

But as for the map view, what's the point if you can't click on a location and see all the photos in that location? I do this all the time on my iPhone--I'm looking for a particular photo, don't know quite when it was but I know where I was. Go to the map view in the photos app, zoom in until I'm only seeing photos in the area I know the photo was taken, tap on that indicator for a bunch of photos, and there I have all those photos to scroll through, just from that location. It's great, I can do it on a location as big as a continent or as small as my house, and it's very quick--takes make 10 seconds to find a specific photo this way.

It makes sense that a map view in a photos app would work this way, and then whatever load in the front and backend that it causes could be worked on.

**Comment by graciousgrey** [comment]
> It makes sense that a map view in a photos app would work this way, and then whatever load in the front and backend that it causes could be worked on.

@fivestones We will find a nice solution when we find time to work on this :)

**Comment by alber70g** [comment]
I have an idea on how to do this:

We could split the circle in a upper and bottom half. 
Upper half: shows a `+` that the user can tap/click and it'll zoom (current behaviour)
Lower half: shows the number of pictures in that area that the user can tap/click that takes the user to a album view of those pictures

**Comment by ttimasdf** [comment]
As for zooming, on desktop we can use mouse scrolls, on mobile device we can use two-finger gestures. Clicking on the circle could be used to browse all the photo inside the whole group. It's rather intuitive.

Here's a screen record from Huawei Gallery app, may be a good reference i think?

https://user-images.githubusercontent.com/2762704/181719030-a8a664c2-dc1d-4972-b37b-be9bc3b437af.mp4

**Comment by fivestones** [comment]
I agree, there is no need for any clicking/tapping on the numbers to be a zoom option. We can just zoom with two fingers or mouse scroll.
I'd love to see it look like in the above Huawei Gallery app.

**Comment by lastzero** [comment]
Huawei uses local images in a native app. We use thumbnails with a responsive Web UI that must be fast and work on all browsers and devices. So it's a tiny bit more complicated.

**Comment by fivestones** [comment]
I suppose it is. But to change from the current, not very useful behavior (zooming in when clicking on a number) to a very useful behavior (showing a search result of all the photos in that area when clicking on the number) should be a super easy change to make. The photos in that area have already been found in the database, which is how we know how many there are. We just need the link to go to a display of all of those particular photos.

If we want to zoom, we can already do that without clicking on the numbers.

To show a photo, just pick the first one out of all those represented by a number, and show it there. Already when you zoom in until only a single photo is on a particular place on the map, that photo is shown. So it shouldn't be hard to show that same number to represent a bunch of photos in a more zoomed out view.

**Comment by heikomat** [comment]
No promises, but maybe I'll try creating a POC for this feature on the weekend

**Comment by heikomat** [comment]
Had a couple hours today. This is the current progress:
- implement that pictures are always clustered, no matter the zoom level
- implement showing the photos of a cluster (not actually 100% exact, but good enough)

Still a lot to do (its rough around the edges, has no tests etc), but it proves that it's doable.

See https://github.com/photoprism/photoprism/pull/2596

https://user-images.githubusercontent.com/20770029/183269072-0e05734c-63d0-4d5a-93d6-10d90d3468c9.mp4

**Comment by svengreb** [comment]
@heikomat That already looks great for a POC, well done! 
I'd like to throw in a small improvement suggestion: Instead of showing only numbers it would be nice to use a photo of this cluster instead as "background". This way it would be easier to quickly find what you are looking for instead of having to click on each cluster in an area when you are not quite sure if it was really taken at this specific location.
The number could be shown either as overlay or maybe we can introduce a new variant implementation of the cluster circle component that can have a small "badge" outside of the circle that contains the number. The second option would be great to prevent contrast issues of the number when the "background" photo has the same color like the font.

To allow users to change this behavior we can make this customizable through new setting keys, e.g. to make it possible to select a specific criteria which image will be used as "background" (latest photo, most viewed etc.) and, of course, disable the feature to only show numbers instead like already possible in your POC.

**Comment by ttimasdf** [comment]
> Instead of showing only numbers it would be nice to use a photo of this cluster instead as "background". This way it would be easier to quickly find what you are looking for instead of having to click on each cluster in an area when you are not quite sure if it was really taken at this specific location. 

@svengreb This idea has already been considered by project developer. So, no hurry 😉 

> Huawei uses local images in a native app. We use thumbnails with a responsive Web UI that must be fast and work on all browsers and devices. So it's a tiny bit more complicated.

**Comment by heikomat** [comment]
All photo IDs of a cluster are known when the cluster is rendered.

Showing **A** photo of a cluster is not a problem I think. Showing a photo based on some criteria would be more difficult.

The map already renders custom elements for single images.
Rendering a different custom element, for example with a counter attached in a little bubble, should be no problem as far as I can tell.

**Comment by heikomat** [comment]
I hope no one waiting for this feature is in a hurry.
I'd really like to spend another night or two developing this, but time is tight in the coming days/weeks.
With some luck I'll have some free time at the end of next week

**Comment by Istria1704** [comment]
I found this thread just before I almost opened a new one about the same thing. Very nice it is being worked on!

I also agree that there should be a way to "show all photos in cluster".
If the zooming function is simple replaced by it, I'd be fine with that. But maybe something like double-tapping is possible to keep both functions? That could work universally on both desktop and mobile, no?

**Comment by heikomat** [comment]
I'm slowly but surely making progress whenever i find some free time.
In it's current state
- the cluster-dialog now has a sticky header with a close-button
- it can be closed via back-button

I tried making the dialog fullscreen on screens with `< 600px` of width, but that somehow breaks closing it?
For now, having it not-fullscreen isn't to bad, even on mobile devices.

a known issue is, that opening a cluster currently resets the search if a search-term was entered.
Other than that, and a lot of testing and a little cleaning up, the only missing thing afaik is implementing better custom-elements to be rendered on the map (something like maybe the first 4 images in a cluster with an outside bubble displaying the image count)

**Comment by heikomat** [comment]
current progress. What do you think about the squared rectangles instead of circles (to see more of the images), and borders and counter bubbles?

<img width="1255" alt="Screenshot 2022-09-17 at 09 41 55" src="https://user-images.githubusercontent.com/20770029/190846319-798d1e38-77a5-412f-8b5d-1bdd0d9cae2a.png">
<img width="1255" alt="Screenshot 2022-09-17 at 09 43 50" src="https://user-images.githubusercontent.com/20770029/190846317-60cb5224-87a8-4171-9205-fef76d7a4499.png">

**Comment by svengreb** [comment]
The count badge/bubble is nice and definitely better that just a one-colored circle without an image preview!
Showing multiple image previews is also a nice idea, but it should be opt-in since the images are really small and maybe a single image might be preferred. Adding this as an option to the settings would be really great.

**Comment by lastzero** [comment]
The new thumbnails on the map look great! As for the overlay, I haven't had enough time to think about it and didn't want to start a discussion based on unfinished work. An alternative could be to display the images at the bottom with horizontal scrolling, similar to Google Maps and other services. Of course, not so many images would be visible there at the same time and it would require developing yet another view.

**Comment by heikomat** [comment]
i'll try to fix the "opening a cluster resets the search" tomorrow or in two days (will have some free time then).
@lastzero do you consider that Bug a show-stopper?
- On the one hand: It is a real bug
- On the other hand: It is not a regression (as this feature didn't exist before)

More important though: Are there any missing must-haves for that feature i should work on before you consider merging it?

**Comment by graciousgrey** [comment]
@heikomat The bug itself would not be a blocker. 
But since we just changed 10 000 lines of code, we want to test these changes thoroughly before merging more changes.

**Comment by heikomat** [comment]
**TL;DR:**
- The last known bug is fixed
- The UI got some small improvements
- clusters with few images look kinda silly but are usable
- I'm basically done, but haven't written any tests
- I assume the next step is to wait for @lastzero and @graciousgrey to find the time to deal with the PR?

**Long version:**

I've got the "opening a cluster resets search"-bug fixed :)
Also added some ui improvements like increasing clusterRadius and element-size, making thumbails cover their parent-element instead of looking zoomed in or not filling up all the space. This at least somewhat improves the "tiny-thunmbnail"-problem with having 4 images in a single element.

The last thing that is not quite how i like it is the dialog-size when opening a cluster with only a few images.
the photo-views use viewport-relative media-queries to set block-parent-relative element sizes.
This is fine as long as the photos-view ist about as large as the viewport (which it always was up to now), but looks broken when the photo-view is much smaller than the viewport. In that case the thumbnails become tiny.

To somewhat mitigate that i make the dialog be full width and at least 80% view-height. Still looks kinda bad with that much empty space, but at least the thumbnails aren't tiny.
If a cluster has a lot of images, it looks good :)

| <img width="1680" alt="Screenshot 2022-10-03 at 00 41 52" src="https://user-images.githubusercontent.com/20770029/193479588-577af8b0-2b52-416e-bb88-d8e0fa81e79b.png"> | <img width="1680" alt="Screenshot 2022-10-03 at 00 41 58" src="https://user-images.githubusercontent.com/20770029/193479592-218f7e65-b132-4de7-8d56-6bf4b8038324.png"> |
| --- | --- |
|  <img width="533" alt="Screenshot 2022-10-03 at 00 55 46" src="https://user-images.githubusercontent.com/20770029/193480001-bd187403-4fb6-4bb5-8651-aec4752f21ad.png"> | <img width="622" alt="Screenshot 2022-10-03 at 00 50 47" src="https://user-images.githubusercontent.com/20770029/193479855-bd5ca8c5-726a-44de-b3f5-a47dc49822aa.png"> |

I'm now at a stage where i would consider the changes done. If someone has further suggestions or knows of bugs i'm open to adjust or add things, but from my point of view there is nothing left that is **required** for a first release.
I'm not yet familiar with the frontend test-setup. Are these changes, that would fit into the existing frontend-tests @graciousgrey?

**Comment by graciousgrey** [comment]
@heikomat  Thank you for all your work :heart: 
We will merge this as soon as possible. I will then add/adapt the  frontend tests :)

**Comment by kvalev** [comment]
@heikomat  Great work, I am eagerly awaiting the next release (hopefully your PR will be merged by then) 👏! Have you considered making a custom component, containing the map and all the customization? I think there are few other ideas (such as this one - #124), which can greatly benefit from a common component.

**Comment by dethpickle** [comment]
I'm glad I found this thread.  This is definitely something that will define usefulness for me.

I do want to ask for a bit more details on what's being accomplished though.  This is all about establishing a filter from the Places map.  Right?

My desired result is around a lot of aerial photos at fixed waypoints over time.  There was a fire in an open area next to my house over the summer and I've been flying a pattern of 17 waypoints every weekend since.  So I have tight little clusters of photos - each one about 20+ photos.  I'm hoping to either click on a cluster or better yet, drag a marquee or rubberband around them and then they would be the filtered set.  Ideally, I'd then be able to further refine it by ordering newest to oldest or oldest to newest as well as putting an absolute date range on it.  Basically, normal multi-part filtering, but on a selection from the Places.

Is that where this is heading?  Right now, I can click on one of the little isolated clusters and it will let me page through the photos (mostly at that location).  But the order isn't chronological and I can't control it.  Furthermore, it eventually bleeds into the surrounding photos.  I'd really like it to be a filter so I could just look at one isolated site from oldest to newest to watch progress of the plants coming back.

This seems like a great project and is likely to become my self-hosting storage for crap-tons of drone photos.  Thanks and well worth the sponsorship just as it is right now.

![Photoclusters2](https://user-images.githubusercontent.com/3790742/209448738-f240c049-7792-493f-86b1-1001a0212380.jpeg)

**Comment by dror3go** [comment]
@dethpickle May I ask why don't you create an album or assign a label/keyword for this known-set of photos, and then browse them instead of keep going to the map and search for a very unique area?

**Comment by dror3go** [comment]
@dethpickle May I ask why don't you create an album or assign a label/keyword for this known-set of photos, and then browse them instead of keep going to the map and search for a very unique area?

**Comment by dethpickle** [comment]
Sure, I could do that.  But my suggestion is around making the map more useful.  It could be a very efficient and intuitive way to filter by gps.  And it'd be way easier than assigning labels or keywords every time I'm importing photos each week.

**Comment by rslite** [comment]
This is also something I would find very useful. I like to tag my pictures by location and then I would like to be able to go to a certain level of zoom and then have a button that would select all the photos included in that level of zoom (for example). Further filters would then apply, as mentioned above. Clicking on a group of pictures is not very useful, since they would have slightly different locations, thus I wouldn't be able to filter all of them.

**Comment by lastzero** [comment]
In order to get this released, I've updated the styles of the map markers to match the colors of the currently enabled theme.

#### Before

![maps-old](https://github.com/photoprism/photoprism/assets/301686/25084f24-6f25-4fe9-a1e0-3407f7350bcf)

#### After

![maps-new](https://github.com/photoprism/photoprism/assets/301686/bdebcbf9-cb9e-4fc9-a9c0-55c711480e42)

#### What else needs to be done?

As a next step, we need to figure out what to do with the overlay and maybe make it optional so that you don't always have to click multiple times to open the pictures in full screen mode, but only when you want to or when there are a certain number of pictures in a cluster.

Also, there are still some bugs to be fixed. For example, when you search for a term in the maps, the results view shows all the pictures in that area, not just the ones that match the search.

**Comment by red-avtovo** [comment]
How deep can we zoom in in order to have clusters?
When we click on cluster do we see the list of the photos as the gallery on a separate screen?

**Comment by lastzero** [comment]
You can test the current implementation on our public demo at https://demo.photoprism.app/. It will be improved based on our UX testing and our demo will be updated regularly so you can test the changes.

**Comment by red-avtovo** [comment]
That looks and works just perfect to me. That is exactly the functionality I was waiting for. Looking forward getting it from the released version. Great job!

**Comment by red-avtovo** [comment]
I found out that the cluster filtering doesn't work as expected
Hope it's a known issue that will be fixed before the release

![Image](https://github.com/photoprism/photoprism/assets/10682826/e46e90a7-5f88-461f-9750-d50dbcf6975c)

**Comment by lastzero** [comment]
Yeah, that's one of the things I was referring to when I said there is still work to do.

**Comment by lastzero** [comment]
It seems that the markers are no longer updating properly anymore, maybe because of the changes added in #3577? The following screenshots were taken on our public demo and should show exactly the same (number of) pictures. Note that in Europe only one picture is shown until you zoom in (and even then the position is often wrong, e.g. a cluster taken in Berlin is shown in Erfurt until you zoom in and out several times):

![Screenshot 2023-08-15 at 07 26 05](https://github.com/photoprism/photoprism/assets/301686/65f88d8e-51bf-4556-aaab-7a90e77a480c)

From what I know, the picture shown in Europe in the screenshot above was actually taken in Canada. When you zoom in, it's gone and the actual pictures taken in Europe are displayed:

![Screenshot 2023-08-15 at 07 26 46](https://github.com/photoprism/photoprism/assets/301686/4794d805-0345-4171-ac6e-52efc6ae3313)

**Comment by reubendowle** [comment]
> It seems that the markers are no longer updating properly anymore, maybe because of the changes added in #3577? The following screenshots were taken on our public demo and should show exactly the same (number of) pictures. Note that in Europe only one picture is shown until you zoom in (and even then the position is often wrong, e.g. a cluster taken in Berlin is shown in Erfurt until you zoom in and out several times):

I am pretty sure that yes, it is the change I did in #3577. It seems to me that it DOES always render correctly, it is just that there is a delay till the last tile gets loaded which in some cases cane be very slow.

I will create a new pull request to revert or fix the portion of that change causing this issue.

**Comment by lastzero** [comment]
@reubendowle Finally seems to work with these changes! We will soon provide an [updated preview build](https://docs.photoprism.app/getting-started/updates/#development-preview) for further [testing](https://demo.photoprism.app/library/places)...

**Comment by lastzero** [comment]
@reubendowle @heikomat Styles could still be improved, but it's generally working now. Sometimes, only a small dot is displayed on the map without any further details like a preview image or the image count. It wasn't noticed before, because it was invisible with the old styles. Seems to depend on the zoom level, with more or less zoom the cluster preview with number of images is shown without any problems:

![cluster-view](https://github.com/photoprism/photoprism/assets/301686/d6b7ca04-dc01-4df5-9e73-e62dd1e23e28)

**Comment by lastzero** [comment]
When you load the map in our demo for the first time, there are sometimes some duplicate markers that disappear when you move the map. This needs to be fixed next.

**Comment by lastzero** [comment]
@reubendowle @heikomat Here is an example of my personal photo library showing duplicate cluster markers after initialization (it later works as expected and adding a move event handler doesn't fix it):

![too-many](https://github.com/photoprism/photoprism/assets/301686/29faec9f-9659-44d1-af3b-c1d309d4e775)

Edit: As a workaround, it is possible to set the animation speed from "None" e.g. to "Fast" in Settings > General.

**Comment by lastzero** [comment]
@reubendowle @heikomat Most of the rendering issues, especially duplicate clusters after initialization, seem to be gone with my latest improvements. I also reduced the size of the clusters slightly and used fewer thumbnails for smaller clusters to minimize lag and cover less of the underlying map on mobile screens:

![clusters](https://github.com/photoprism/photoprism/assets/301686/a32e0b10-8126-46d7-92aa-ada5557b10d8)

Next, I'll take a look at the dialog that opens when you click on a cluster. My idea is to display it as an overlay at the bottom instead, so that it doesn't cover the whole map but only a part of it, e.g. with this component or something similar:
- https://v15.vuetifyjs.com/en/components/bottom-sheets

**Comment by red-avtovo** [comment]
How much space do you plan to use?
Right now it is nicely displayed even if I open the filtering section. I still see some images and I understand that I stay on the same page. If filters take over the whole segment, then some people could be confused without seeing photos at the same place. 

Hope it makes sense

**Comment by lastzero** [comment]
> How much space do you plan to use?

It should be like a drawer that you can expand and collapse as needed.

Unfortunately, the clusters still don't work properly, as I just noticed when [testing it on our public demo](https://demo.photoprism.app/library/places). The images from Germany are displayed in Canada, which explains why there is only a dot as a placeholder visible in Germany (each cluster has a unique ID and can therefore only be displayed once):

![Screenshot 2023-08-17 at 15 39 47](https://github.com/photoprism/photoprism/assets/301686/849d919d-a74c-4338-ac60-60718c293ac0)

All in all, it looks like there's still a lot of work ahead, so I can't say when it will be ready for release :(

**Comment by red-avtovo** [comment]
No rush. We were waiting it for so long so we will be happy to get it when it is fully ready.
Thank you for driving it ❤️

**Comment by lastzero** [comment]
Maybe the problem is with the cluster IDs, if they are not as unique as previously thought. Or there is a bug in the library so that Canada and Germany get confused? 🤔

**Comment by reubendowle** [comment]
> Maybe the problem is with the cluster IDs, if they are not as unique as previously thought. Or there is a bug in the library so that Canada and Germany get confused? 🤔

I put a bunch of debug prints into various points in the code and when I was able to reproduce similar behavior on my local photo library I realized that this seems to happen when multiple updateMarkers() calls are running concurrently. I think with the changes in my latest pull request to serialize the updates this issue is fixed.

I also noticed that the calculated list of clusterIDs often has duplicate entries. I think maybe this is due to the map wrapping around? Anyway, I have also removed these duplicates so we don't end up creating the same cluster twice. This change has no effect that I could see on the rendered clusters, but might reduce some duplicated calculations.

**Comment by lastzero** [comment]
@reubendowle Thank you! From how the cluster update code looks, one problem could be that it is a "two-pass" approach where the features are queried multiple times, i.e. first the cluster ids and then the related images via `getMultipleClusterFeatures()`, so the data could have changed in between?

**Comment by lastzero** [comment]
@reubendowle I've started a new preview build so we can [test your changes on our public demo](https://demo.photoprism.app/library/places). It's easy to see there if the number of pictures in the clusters adds up when zooming in and out! 🤓

One thing I noticed is that when I zoom or move the map, the clusters don't update simultaneously and even after I stop doing anything, it can take a while for the new clusters to be visible and the old ones to disappear. With your changes, would it help to also listen for the `move` event instead of just `moveend`? Based on my tests, it hasn't made any difference so far (just caused more computational overhead).

**It seems worth looking for ways to further improve the performance of the code, as the cluster update function is called very often!?** :detective:

**Comment by lastzero** [comment]
@reubendowle Unfortunately, despite your improvements, our demo [still shows the pictures in the wrong location](https://github.com/photoprism/photoprism/issues/1187#issuecomment-1682322059) :cry:

![changes](https://github.com/photoprism/photoprism/assets/301686/87d75284-651f-4bdd-8dd1-0f16afbed183)

*Note that this depends on the zoom level and maybe the browser window size. It doesn't always happen.*

@reubendowle @heikomat I haven't gone through it with a debugger yet, but as [written above](https://github.com/photoprism/photoprism/issues/1187#issuecomment-1683380897), the problem could be caused by a "two-pass" approach where the features are queried multiple times, i.e. first the cluster IDs and then the related pictures via `getMultipleClusterFeatures()`, so the data might have changed in between (or is outdated when the function returns)?

**Comment by reubendowle** [comment]
It is really hard for me to reproduce this on my local data. The demo seems to do it reliably. I put a breakpoint in the demo instance when just the cluster that is supposed to be in germany is being rendered.

![image](https://github.com/photoprism/photoprism/assets/9706673/8f91b628-d4fb-4643-8526-6bd98ba833df)

It is really strange - the coordinates for the cluster are neither in USA or Germany. Seems to be in Antarctica I think? So it seems the data being returned for the clusters is somehow messed up.

**Comment by lastzero** [comment]
@reubendowle You can download the files used in our demo for local testing: https://dl.photoprism.app/qa/demo.tar.gz

**Comment by reubendowle** [comment]
> @reubendowle You can download the files used in our demo for local testing: https://dl.photoprism.app/qa/demo.tar.gz

I downloaded the data, and did some testing. I was never able to reproduce the same problem as was seen three days ago. However I was able to consistently see a similar issue with a cluster from USA rendering in Tubingen when I zoomed in there.

I think this variation might be because the data I downloaded (160 photos) does not exactly match what is on the public demo (136 photos).

**Comment by lastzero** [comment]
@reubendowle You are right! The demo now probably uses the files in `plus.tar.gz` as we added vector graphics to test the optional functionality in our plus version: https://dl.photoprism.app/qa/

**Comment by lastzero** [comment]
With these fixes, all markers - clusters and photos - finally work in my local environment and on our demo. I have started a new preview build so we can do more testing. It should be available within the next hour:

- https://docs.photoprism.app/getting-started/updates/#development-preview
- https://docs.photoprism.app/release-notes/#development-preview

**Comment by lastzero** [comment]
Here is an updated screenshot of the cluster view (a new [preview build](https://docs.photoprism.app/getting-started/updates/#development-preview) will also be available for testing soon):

![cluster-view](https://github.com/photoprism/photoprism/assets/301686/aebc240d-64f0-4086-859f-3e7b8ab9c794)

Note that you can now continue using the map while the results view is open. The photos and clusters on the map are also displayed and updated correctly.

**Comment by lastzero** [comment]
Same view with a light theme (Lavender):

![cluster-view](https://github.com/photoprism/photoprism/assets/301686/679f5798-89d9-4890-8a33-6fcaa958f060)

**Comment by heikomat** [comment]
This refactor looks pretty cool, well done @lastzero!

**Comment by lastzero** [comment]
We'd love to release this on Friday! Everything looks good to me now, except for a [missing resize handle](https://github.com/photoprism/photoprism/issues/3747) so you can adjust the size of the cluster preview:

![resize-handle](https://github.com/photoprism/photoprism/assets/301686/976fcc87-07bd-425c-aabe-936d052d5af3)

That being said: If you want to help with testing, now would be a great time! :partying_face:

**Comment by lastzero** [comment]
Final screenshot with the [map style selector](https://github.com/photoprism/photoprism/issues/2106) and places cluster view open (the [distance scale](https://github.com/photoprism/photoprism/issues/2106#issuecomment-1728124244), also new, is hidden behind the overlay in this case, but I don't think that's a good enough reason to move it to the top):

![places-cluster-view](https://github.com/photoprism/photoprism/assets/301686/cd48e79f-b2a8-4fb5-bc54-52467b15b743)
